grpc/estimator.proto

```diff
@@ -105,42 +119,54 @@ message SettingsResponse {
 // TODO(etom): we want a breakdown of the cost into the different sensors and
 // forces
 
-message Cost {
-  optional double total = 1;
-  optional double prior = 2;
-  optional double sensor = 3;
-  optional double force = 4;
-  optional double initial = 5;
+message CostRequest {
+  optional bool derivatives = 1;
+  optional bool internals = 2;
 }
 
-message CostRequest {}
-
 message CostResponse {
-  Cost cost = 1;
+  double total = 1;
+  double prior = 2;
+  double sensor = 3;
+  double force = 4;
+  double initial = 5;
+  repeated double gradient = 6 [packed = true];
+  repeated double hessian = 7 [packed = true]; 
+  repeated double residual_prior = 8 [packed = true];
+  repeated double residual_sensor = 9 [packed = true];
+  repeated double residual_force = 10 [packed = true];
+  repeated double jacobian_prior = 11 [packed = true];
+  repeated double jacobian_sensor = 12 [packed = true];
+  repeated double jacobian_force = 13 [packed = true];
+  repeated double norm_gradient_sensor = 14 [packed = true];
+  repeated double norm_gradient_force = 15 [packed = true];
+  repeated double prior_matrix = 16 [packed = true];
+  repeated double norm_hessian_sensor = 17 [packed = true];
+  repeated double norm_hessian_force = 18 [packed = true];
+  int32 nvar = 19;
+  int32 nsensor = 20;
+  int32 nforce = 21;
 }
 
 // TODO(etom): all the protos below use a dict of arrays, but they should use an
 // array of dicts.
 
-message Weight {
-  optional double prior = 1;
+message Noise {
+  repeated double process = 1 [packed = true];
   repeated double sensor = 2 [packed = true];
-  repeated double force = 3 [packed = true];
 }
 
-message WeightsRequest {
-  Weight weight = 1;
+message NoiseRequest {
+  Noise noise = 1;
```

Comment:
Given your scheme of using the same request for get and set, I would make this argument optional, so it's possible to get the Noise value without changing it. Same for other requests.